### ISSUE TITLE
Rewriting repo history to remove 'target/' dir could significantly improve repo size

### ISSUE THREAD

**Issue by kbknapp** [GyulyVGC/sniffnet#172]
TL;DR: Cleaning the `target/` dir from some previous commits reduces the total repo size from ~210MB to ~110MB.

I was on a slow connection when cloning this repo and I realized it was decently large for a new repo:

```
$ du -hxs sniffnet/
209MB    sniffnet/
```

I then ran a command to see the largest 10 files in the history (expecting resources or fonts):

```
$ git rev-list --objects --all \
  | git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' \
  | sed -n 's/^blob //p' \
  | sort --numeric-sort --key=2 \
  | cut -c 1-12,41- \
  | $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest \
  | tail -n10

10ca1380679d  9.9MiB packet_sniffer/target/debug/deps/libsyn-1b755dd9a1cd0131.rlib
54d8337b22c7   12MiB packet_sniffer/target/debug/deps/libregex_syntax-9e6c6888eea69ccd.rlib
0995ee2a7d96   13MiB resources/fonts/full/sarasa-mono-sc-bold.ttf
67ebe0547390   13MiB resources/fonts/full/sarasa-mono-sc-regular.ttf
26238878ca0e   14MiB resources/fonts/full/sarasa-mono-sc-bold.ttf
9994b9f9c3da   14MiB resources/fonts/full/sarasa-mono-sc-regular.ttf
8ae5e938730b   14MiB resources/fonts/full/sarasa-mono-sc-regular.ttf
330ef6743ac5   15MiB packet_sniffer/target/debug/deps/libclap-6006cabbd19c8508.rlib
26ac37a46cf3   15MiB packet_sniffer/target/debug/deps/libclap-57046093200aadfb.rlib
557fd1a7a5f6   15MiB packet_sniffer/target/debug/deps/libsyn-97bf2b04b67d3019.rlib
```

I noticed some `target/` references in there and realized this repo could probably benefit from rewriting the history to remove those objects from history.

So as a test, I ran the following commands using [`bfg`](https://rtyley.github.io/bfg-repo-cleaner/):

```
sniffnet$ bfg --delete-folders target --no-blob-protection  .
[.. lots of output .. ]

sniffnet$ git reflog expire --expire=now --all && git gc --prune=now --aggressive
```

These are the results aftwards:

```
$ du -hxs sniffnet/
112M    sniffnet/
```

At this point you'd have to force push to `main` to actually keep these results.

**Comment by GyulyVGC** [comment]
Will this have some consequences when I'll merge one of the existing secondary branches to main?
I mean, I would have to re-do it afterwards right?

**Comment by kbknapp** [comment]
Yeah it's just any other force push to `main`, you'd need to rebase any in-flight work/PRs onto the `main` after doing the cleanup. There shouldn't be any conflicts because the cleanup changes only involve removing `target/`; which I'm assuming none of the in-flight work/PRs are doing anything with?

You can also test this locally to ensure there are no issue though, prior to force pushing anything. Actually, you can even do this on a fresh clone of the repo if you don't want to risk your main working directory just to ensure if anything bad happens locally you delete the fresh clone.

```
$ [ .. perform steps listed above .. ]
$ git checkout my-in-flight-pr
$ git rebase main
```
Assuming no conflicts, you just repeat this for every in-flight branch/PR.

At this point, if there are conflicts you can "abort"

```
$ git rebase --abort
```

Assuming no issues you'd just force push `main` and also force push all the PR branches to their respective remote/branch.

One other item that feels strange, but is actually not as scary as it sounds: if you have any clones who's `main` points to the "old" `main` doing a `git pull` would result in a bunch of conflicts/errors. The solution would be to just delete the _local_ `main` branch and fetch/pull the new remote `main`. e.g.

```
$ git branch -D main
$ git fetch
$ git switch main
```

This is why people are usually so against force pushing, it causes a bunch of work 😄 However, in instances like this, where you're saving 100MB+ (or say redacting secrets that were accidentally committed) I personally think it's justified. But that's my subjective opinion.

Granted I know `git` can be opaque and it can make people nervous when you have in-flight work and you're messing with `main`. If this all feels too risky no worries, I wouldn't be offended or anything if you closed the issue to address some other time or just accept size cost 😄

**Comment by GyulyVGC** [comment]
So... 
- I've merged the secondary branches to the (old) main to avoid conflicts
- I've locally rewritten main history 
- But.. when trying to rebase a secondary branch to the (new, local) main, I get a bunch of conflicts despite the secondary branches are not behind main (git merge main -> already up to date, but git rebase main -> conflicts)

**Comment by kbknapp** [comment]
If I'm understanding the actions you took, you shouldn't need to rebase the secondary branches into the new main, if they were already merged into the old main prior to rewriting the history. That will indeed create conflicts.

In order to keep the secondary branches you only need to do one of the options below:

- Option 1
  - Rewrite local main history
  - (Optionally) rebase a secondary branch onto the new local main just as a test before force pushing
  - Force push the new local main to the remote
  - (If you did not do the secondary-branch rebase already as a test) Rebase the secondary branch onto main 
  - Force push the secondary branch to it's remote
  - Merge the secondary branch like normal
- Option 2
  - Merge the secondary branches into old main
  - Rewrite local main history
  - Force push main

With option two (which is what it sounds like you did?) there is no need to rebase the secondary branches later because they were already merged into main before rewriting the history.

With option one, that first rebase is just a test to make sure there will be no conflicts before you force push main.

**Comment by GyulyVGC** [comment]
Yep, option 2 is exactly what I did.
I have now the secondary branches with 600+ commit behind and 600+ commits after main.
I guess I just have to pull and push them, right?

**Comment by GyulyVGC** [comment]
I'm also wondering: if I push a big file to git and later remove that file because not needed anymore, will git keep references (and waste lots of space) forever? Is rewriting history the only solution to this problem?

**Comment by kbknapp** [comment]
Replying in reverse order 😄 

> I'm also wondering: if I push a big file to git and later remove that file because not needed anymore, will git keep references (and waste lots of space) forever? Is rewriting history the only solution to this problem?

Yep that's exactly what happens (and the reason I opened this issue). It's unfortunately, but also understandable why it works this way. There are also some unique circumstances where a large file can even get duplicated in history taking up _even more_ space forever, but I can't remember the exact details on when or why this happens.

> I have now the secondary branches with 600+ commit behind and 600+ commits after main.

If the secondary branch was merged into old main before you re-wrote the history, you actually don't need the secondary branch any longer.

To illustrate (I'm a visual person so images help me); assume we have a repo with 4 commits (`A`, `B`, `C`, and `D`) which is what the `main` branch currently looks like. We also have a `secondary` branch with two additional commits (`E` and `F`). Finally, let's assume commit `B` was the one were we accidentally committed a large file.

> **Note**
> Remember branches are kinda just like pointers to a specific commit.

> **Warning**
> Because in git, all commits must point to the their parent, we can't simply "only rewrite `B`", doing so requires that all child commits also be re-written.

Let me quickly go over Option 1 above for completeness sake.

![git1](https://user-images.githubusercontent.com/6942134/236709557-df9bb1af-8204-4f8b-90c4-5c6278af29c6.png)

Here we start with the repo as described above. Then we re-write history resulting in 3 commits on `main` being re-written as `B2`, `C2` and `D2`.

> **Note**
> Although in the illustration it's easy to see `A` is being re-written as `A2`, in real git there isn't such an easily correlation because `A` would actually be some SHA2 hash and `A2` would be some entirely different SHA2 hash. 

Notice though that before rebasing, `secondary` still points to the "old" set of commits. Hence we "re-base" or change the base commit from `D` (old main) to `D2` (new main).

----

Switching to Option 2:

![git2](https://user-images.githubusercontent.com/6942134/236710049-1ff5435d-5276-4c09-99b0-10b28b576d15.png)

We start with the same repo set up.

But next we merge `secondary` into `main` which results in a merge commit `M`. At this point normally we'd delete the `secondary` branch (remember it's just a pointer, we aren't actually deleting the commits).

Now we re-write history which results in a new set of commits `B2`, `C2`, `D2`, `E2`, `F2`, and `M2`.

Now if you still have `secondary` sitting around in git metadata you'd probably see something like `5 commits ahead, 5 commits behind of main` which just saying, "I have 5 commits `main` doesn't have (`B`, `C`, `D`, `E`, and `F`), and `main` has 5 commits that I don't have (`B2`, `C2`, `D2`, `E2`, and `F2`" which actually doesn't matter because `main` as we know it _already_ has the `secondary` branch commits we wanted from `E2` and `F2` resulting in a merge commit `M2`.
---

One final thing to keep in mind here that isn't depicted in the illustrations is that each "remote" is it's own island/view of the world. So while the illustrations were all working on your local copy (which is still a "remote" so to speak), you're not affecting the github remote (usually `origin`) or any of the PR remotes (assuming they're submitted by other users). All those other remotes are still only tracking what `main` looked like _before_. And once you force push `main` after re-writing history, all those remotes will likely show the same `+600,-600` commits, or something similar until they effectively delete _their_ `main` and re-pull the new main and rebase any non-merged work-in-progress.

I hope this helps! 😄

**Comment by GyulyVGC** [comment]
For the moment I'd prefer avoiding the procedure considering that the app will be updated soon and several contributors with an already existing fork will update the translation in their respective language.

Thank you very much for these insights however. You made me aware of problems I was not considering at all. I will come back on this issue for sure.
